### ISSUE TITLE
Remove time limit for "Press twice quickly" or "Press three times quickly" commands

### ISSUE THREAD

**Issue by Qchristensen** [nvaccess/nvda#11929]
### Is your feature request related to a problem? Please describe.

Many NVDA commands have additional functionality when they are pressed twice quickly.  E.G.:

- NVDA+f12 pressed once reads the time, and pressed twice quickly reads the date.
- NVDA+t pressed once reads the titlebar, pressed twice quickly spells it, and pressed three times quickly, copies it to the clipboard.

There are numerous other examples.

Pressing keystrokes twice or three times for this extra functionality needs to be done within about a second.  This is difficult for some users, and particularly difficult if using Sticky Keys (it can be done, but if you can press and release NVDA then t, then NVDA then t, then NVDA then t all in one second, you probably don't need to use sticky keys to begin with).

Although WCAG is aimed at web applications, the principles can be applied to desktop applications.  Section 2.2 of WCAG deals with timing and basically, where possible, discourages events which have are time limited.  Aside from the way the keystrokes are currently implemented, there is no logical reason that the events themselves, such as copying to the clipboard, spelling phonetically, or reading the date need to be time limited.

### Describe the solution you'd like

If NVDA could remember the last keystroke pressed and where the same NVDA keystroke is pressed twice (or three times), treat it as that keystroke would currently be treated if pressed twice (or three times) in quick succession.

In all cases I can think of, pressing a keystroke twice results in reading or spelling information.  While not intended when trying to press the keystroke three times, this should not cause problems in the middle of the action.

The only time I can think of this would be undesirable is where someone wanted to press the same keystroke multiple times (not quickly) where for instance, the text to be read changed, such as a live region on a web page.  To work around this, the option could be either toggled via a setting, or could be tied to whether sticky keys was enabled.

### Describe alternatives you've considered

The only alternative I can think of would be to assign different keystrokes to the "second" or "third" functions (when you press a keystroke twice or three times), but this would be a very inefficient use of keystrokes, and would be an extra cognitive load on users (it's easy to remember currently that for many NVDA commands, if you want to spell the output, press it twice quickly.

**Comment by CyrilleB79** [comment]
> The only time I can think of this would be undesirable is where someone wanted to press the same keystroke multiple times (not quickly) where for instance, the text to be read changed, such as a live region on a web page.

I agree with that. This use case is frequent.

> To work around this, the option could be either toggled via a setting, or could be tied to whether sticky keys was enabled.

I have had a look at the code; both solutions may be implemented quite easily.

The first solution (make it a setting) may allow a more flexible way to use it. For example it could be an edit field or a combo-box in the keyboard setting panel allowing
to configure the delay between two key press to be considered a double key press. This delay would default to 0.5 (as today) and could also be infinite.

The second solution has pros and cons:
  * does not allow the flexibility of the first solution: the feature is on or off based on the sticky key flag (con)
  * is immediately working for people using sticky keys. It does not force them to know that a setting exists for them and to search for it (pro)
  * is limited to sticky key users whereas other user may want to benefit from this possibility (con)

I am not sure that having an infinite time would be very useful. I hardly imagine someone needs more than 30 seconds between one key press and the following one. Moreover, the maximum delay between a double key press should not exceed the mean time required for someone to forget he has already pressed the key.
An edge case (maybe a bit exagerated) illustrating this would be someone who wants to reset his configuration to last saved configuration with the NVDA+control+R shortcut:
* 1st press: configuration is reset to saved configuration
* waiting more than 30 seconds, the user does not recall very well if he has already reset his configuration
* 2nd press: Nothing is reported. The user think he has not pressed very well the keystroke
* 3rd press: the configuration is restored to factory configuration

I am not a user of sticky key and do not know anyone in this case. Thinking only at my own use, I am happy with the current way handles multiple key press. Thus a feedback from sticky key users or from people having difficulty with multiple key press would be welcome.

**Comment by DrSooom** [comment]
See also the expected behavior in issue #8517, as the basic problem is quite similar.

**Comment by CyrilleB79** [comment]
> See also the expected behavior in issue #8517, as the basic problem is quite similar.

Not so much to do with #8517, unless on the UI item, provided we agree on the possibility to set the double-press delay via the UI. Let's comment further there if needed.

**Comment by DrSooom** [comment]
@CyrilleB79: Yepp, I exactly meant the GUI elements here.

**Comment by leonardder** [comment]
I think I'd like this suggestion. Should be possible to implement similar to the braille message timeout, with a spinner in miliseconds.

**Comment by bhavyashah** [comment]
Just want to second @CyrilleB79 on "This use case is frequent." I certainly see why this change is necessary for users with dexterity-related impairments but I would request that the solution be through an added option not an unconfigurable alteration.

**Comment by CyrilleB79** [comment]
Yes, no problem for making something configurable. I have already a branch to fix this issue. But I still have to fix a little issue (probably related to wx control) before submitting the PR for this.

**Comment by CyrilleB79** [comment]
For people interested in this feature, please look at https://github.com/nvaccess/nvda/pull/12049#issuecomment-1185605016 and comment there for possible implementation in the GUI to fix this issue.
Thanks.
